sql/sem,sql/parser: fix the AST anonymization

knz opened the pull request:
Fixes #60703. 

Previously, the tests for AST anonymization were restricted to a few
hand-picked statements in
`sql/sem/tree/format_test.go`. Unfortunately, most AST pretty-printing
is instead tested in `sql/parser` so folk have routinely forgotten to
implement anonymization until now.

This patch attempts to fix this by using a datadriven replacement
to `TestParse` that also checks the anonymization code path.

Release note (bug fix): More data from SQL statements is anonymized in
telemetry updates and crash reports.

knz committed:
sql/sem,sql/parser: fix the AST anonymization

Previously, the tests for AST anonymization were restricted to a few
hand-picked statements in
`sql/sem/tree/format_test.go`. Unfortunately, most AST pretty-printing
is instead tested in `sql/parser` so folk have routinely forgotten to
implement anonymization until now.

This patch attempts to fix this by using a datadriven replacement
to `TestParse` that also checks the anonymization code path.

Release note (bug fix): More data from SQL statements is anonymized in
telemetry updates and crash reports.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60676)
<!-- Reviewable:end -->

knz commented:
I am planning to extend this PR as follows:

- fold TestParse2 into the same datadriven test
- use multiple datadriven input files, to split the statements into categories
- move the special cases in `tree/format_test.go` into the new parse tests

dt commented:
This makes me wonder if the type signatures could make the compiler help us out here, e.g. if `Expr.Format(FmtCtx)` is always wrong and we should instead be using `FmtCtx.FormatNode(Expr)` maybe we should rename and/or unexport the former? as it is, it seems like a very easy mistake to make.

knz reviewed (commented):
they are both in the same package, so unexporting doesn't help. 

But we could play with the types yes, so that the wrong call would need a long cast with an identifiers like `ctx.(areYouSureyouDontMeanFmtNodeHere)`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MTkKRvN2R1p8Mf_9-_f:b-k3i4rz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @otan, @rafiss, and @solongordon)



<!-- Sent from Reviewable.io -->

dt commented:
I also wonder if we should split this up: one PR that _fixes_ all the current mistaken usages and can be backported for existing releases that are leaking unscrubbed crash reports or telemetry and then a follow-up with the capital improvements to prevent it happening again?

knz commented:
yes I agree on this split

knz commented:
Ok this is ready for review now. Thanks in advance.

rafiss reviewed (changes requested):
thank you for your extreme diligence! i had some questions about what counts as an identifier. and thank you for filing the follow up issues for JSON and types.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MTmKukg-XQZLXK3-xBU:b86bvqh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, @otan, and @solongordon)

___
*[pkg/sql/parser/testdata/parse/alter_database, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmFVvN8V_Bus0Qhp-O:-MTmFVvN8V_Bus0Qhp-P:b-6kt5cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_database#L63)):*
> ```
> ALTER DATABASE db CONFIGURE ZONE = ('foo') -- fully parenthetized
> ALTER DATABASE db CONFIGURE ZONE = _ -- literals removed
> ALTER DATABASE _ CONFIGURE ZONE = 'foo' -- identifiers removed
> ```

what is 'foo' in this context? not an identifier?
___
*[pkg/sql/parser/testdata/parse/alter_user, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGCkE4KVMgiiSNUFy:-MTmGCkE4KVMgiiSNUFz:b-8ljhfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_user#L25)):*
> ```
> ALTER USER ('foo') WITH PASSWORD (NULL) -- fully parenthetized
> ALTER USER _ WITH PASSWORD _ -- literals removed
> ALTER USER 'foo' WITH PASSWORD NULL -- identifiers removed
> ```

is it intended that the user name is not considered an identifier?
___
*[pkg/sql/parser/testdata/parse/alter_user, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGKdP9WEbYej7R8jX:-MTmGKdP9WEbYej7R8jY:b-lp9w22) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_user#L33)):*
> ```
> ALTER ROLE ('foo') WITH CREATEDB -- fully parenthetized
> ALTER ROLE _ WITH CREATEDB -- literals removed
> ALTER ROLE 'foo' WITH CREATEDB -- identifiers removed
> ```

same for role name?
___
*[pkg/sql/parser/testdata/parse/backup_restore, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGwDp1j889QDU-XRe:-MTmGwDp1j889QDU-XRf:b8lh1fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/backup_restore#L55)):*
> ```
> BACKUP TABLE (foo) INTO ('subdir') IN ('bar') -- fully parenthetized
> BACKUP TABLE foo INTO _ IN _ -- literals removed
> BACKUP TABLE _ INTO 'subdir' IN 'bar' -- identifiers removed
> ```

is directory name an identifier?

what is `bar` in this statement? i didn't see it on this release's docs so i'm guessing a new feature? https://www.cockroachlabs.com/docs/stable/backup.html
___
*[pkg/sql/parser/testdata/parse/copy, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmI9h-7M0AUn__Mm8E:-MTmI9h-7M0AUn__Mm8F:bcjljba) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/copy#L23)):*
> ```
> COPY crdb_internal.file_upload FROM STDIN WITH destination = ('filename') -- fully parenthetized
> COPY crdb_internal.file_upload FROM STDIN WITH destination = _ -- literals removed
> COPY _._ FROM STDIN WITH destination = 'filename' -- identifiers removed
> ```

should destination be an identifier?
___
*[pkg/sql/parser/testdata/parse/create_user, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmIzq3-UrrYYfDyclW:-MTmIzq3-UrrYYfDyclX:begy123) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/create_user#L24)):*
> ```
> CREATE ROLE ('foo') -- fully parenthetized
> CREATE ROLE _ -- literals removed
> CREATE ROLE 'foo' -- identifiers removed
> ```

should role/user names be treated as identifiers?
___
*[pkg/sql/parser/testdata/parse/import_export, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmJIkQ6e9T0tyhjEt_:-MTmJIkQ6e9T0tyhjEta:bp9cb8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/import_export#L15)):*
> ```
> EXPLAIN IMPORT TABLE foo CREATE USING ('nodelocal://0/some/file') CSV DATA (('path/to/some/file'), ($1)) WITH temp = ('path/to/temp') -- fully parenthetized
> EXPLAIN IMPORT TABLE foo CREATE USING _ CSV DATA (_, $1) WITH temp = _ -- literals removed
> EXPLAIN IMPORT TABLE _ CREATE USING 'nodelocal://0/some/file' CSV DATA ('path/to/some/file', $1) WITH _ = 'path/to/temp' -- identifiers removed
> ```

file and directory names?
___
*[pkg/sql/parser/testdata/parse/predefined_functions, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmJcg-96ioAuPL1Whe:-MTmJcg-96ioAuPL1Whf:bgw06cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/predefined_functions#L167)):*
> ```
> SELECT (substring(('f(oabaroob'), (e'\\(o(.)b'))) -- fully parenthetized
> SELECT substring(_, _) -- literals removed
> SELECT substring('f(oabaroob', e'\\(o(.)b') -- identifiers removed
> ```

hmm interesting. i guess i see why this might be fine, but showing the strings still feels like a PII risk


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I see how you are confused by the fact that FmtAnonymize does not redact literal values out, whereas literal values *in general* also contain PII and other sensitive information.

I just did the additional homework and confirm that the data sent out to CRL is always processed by both FmtHideConstants and FmtAnonymize, albeit the fact both are applied is a bit hard to figure out.

Here's the deal:

1. crash reports use `sql.anonymizeStmtAndConstants()` which applies both directly
2. app stats construct a key via `sql.anonymizeStmt()` which only applies FmtHideConstants. That's a feature: we want two similar-looking statements that affect different tables to be bucketed separately, but two statements that affect the same table with different literals to be bucketed together.
3. telemetry reports take the app stat keys (already processed with FmtHideConstants previously), **re-parse them** and then apply FmtAnonymize on top.

It's all a bit confusing. However it should satisfy you in your comments below that things like file paths etc are properly erased in telemetry.

Also it makes me notice there's a bug: we haven't validated that statements that underwent FmtHideConstants can be re-parsed, so we have potentially a bunch of statements that fail their re-parse at step 3 above and are never reported to telemetry. We will need to quantify this. I'm sure @vy-ton will want to know this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MToIic16UV6RSHk1HDy:b-8v0qbt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @otan, @rafiss, and @solongordon)

___
*[pkg/sql/parser/testdata/parse/alter_database, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmFVvN8V_Bus0Qhp-O:-MToIK3NEmz2JSo0nM2h:b-rq52g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_database#L63)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what is 'foo' in this context? not an identifier?
</blockquote></details>

it's a YAML blob. It contains references to the db's regions and localities.
___
*[pkg/sql/parser/testdata/parse/alter_user, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGCkE4KVMgiiSNUFy:-MToIOtECaQWyNDpsvOc:baq8avt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_user#L25)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is it intended that the user name is not considered an identifier?
</blockquote></details>

It's not implemented as one. 
___
*[pkg/sql/parser/testdata/parse/alter_user, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGKdP9WEbYej7R8jX:-MToIRgpDaz-gcH__GXy:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/alter_user#L33)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

same for role name?
</blockquote></details>

ditto
___
*[pkg/sql/parser/testdata/parse/backup_restore, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmGwDp1j889QDU-XRe:-MToIShG7fY9B4TX4xfl:b-gftq3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/backup_restore#L55)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is directory name an identifier?

what is `bar` in this statement? i didn't see it on this release's docs so i'm guessing a new feature? https://www.cockroachlabs.com/docs/stable/backup.html
</blockquote></details>

It's not an identifier, it's a string

I don't know what bar refers to
___
*[pkg/sql/parser/testdata/parse/copy, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmI9h-7M0AUn__Mm8E:-MToIWs36HMs8RkRHdzk:b-o8voit) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/copy#L23)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should destination be an identifier?
</blockquote></details>

generally, anything that refers to a file is implemented as a string / literal string
___
*[pkg/sql/parser/testdata/parse/create_user, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmIzq3-UrrYYfDyclW:-MToIaAs7uEZS2NGCIif:byr4ffv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/create_user#L24)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should role/user names be treated as identifiers?
</blockquote></details>

see above
___
*[pkg/sql/parser/testdata/parse/import_export, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmJIkQ6e9T0tyhjEt_:-MToIbWvBY4eyAQdZ_Oo:b-bagbpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/import_export#L15)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

file and directory names?
</blockquote></details>

again anything outside of db is typically a string
___
*[pkg/sql/parser/testdata/parse/predefined_functions, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTmJcg-96ioAuPL1Whe:-MToIgOhDW_DyWazDFEv:bucq9mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/testdata/parse/predefined_functions#L167)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hmm interesting. i guess i see why this might be fine, but showing the strings still feels like a PII risk
</blockquote></details>

see my comment at top


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Filed #60722 to track this latter bug.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MToSjFS3p19dGERA5fm:b-mze6qh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @otan, @rafiss, and @solongordon)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MTr81U9B59hsdIpuk9w:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @knz, @otan, @rafiss, and @solongordon)

___
*[pkg/sql/parser/parse_test.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTr6wPTDJmfD10BF6K4:-MTr6wPTDJmfD10BF6K5:b-2ldb2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/parse_test.go#L34)):*
> ```Go
> 
> // TestParseDataDriven verifies that we can parse the supplied SQL and regenerate the SQL
> // string from the syntax tree.  If the supplied SQL generates a different string
> ```

[nit] remove the last sentence
___
*[pkg/sql/parser/parse_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTr7ibG-E1ew3EA8z4S:-MTr7ibG-E1ew3EA8z4T:bycpkzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/parse_test.go#L36)):*
> ```Go
> // string from the syntax tree.  If the supplied SQL generates a different string
> // from the sytnax tree, use TestParse2 below.
> func TestParseDatadriven(t *testing.T) {
> ```

[nit] This test could now be unified with `TestParseErrors` (and we could then keep all test files in the same directory `sql/parser/testdata/`). It would also give us the freedom to move error cases in the specific statement files instead of a separate file.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thank you for the explanation. i'm glad the telemetry redacts literals, and nice find with the followup issue you filed

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @knz, @otan, @rafiss, and @solongordon)

___
*[pkg/sql/sem/tree/format.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MUU7B206qwtPlYVI58e:-MUU7B210cL2Ns8DZ70K:b2f5f2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/sem/tree/format.go#L144)):*
> ```Go
> // PasswordSubstitution is the string that replaces
> // passwords unless FmtShowPasswords is specified.
> const PasswordSubstitution = "*****"
> ```

If we make this `const PasswordSubstitution = "'*****'"` then we could solve the first problem here: https://github.com/cockroachdb/cockroach/issues/60722

is that do-able?


<!-- Sent from Reviewable.io -->

rafiss commented:
@knz friendly ping on this -- is still in the works for 21.1?

rafiss commented:
@knz friendly ping now that you are back from vacation. i'm curious to know the plans for completing this.

knz commented:
yeah let's work on a rebase

knz commented:
> If we make this const PasswordSubstitution = "'*****'" then we could solve the first problem here: #60722is that do-able?

that is a very nice idea. Probably worth a followup PR though.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @knz, @otan, @rafiss, and @solongordon)

___
*[pkg/sql/parser/parse_test.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTr6wPTDJmfD10BF6K4:-MVwGHzZ44p27ahEtSbJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/parse_test.go#L34)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] remove the last sentence
</blockquote></details>

Done.
___
*[pkg/sql/parser/parse_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTr7ibG-E1ew3EA8z4S:-MVwGIiR9OFz_fAAOTCQ:bisusra) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/parse_test.go#L36)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This test could now be unified with `TestParseErrors` (and we could then keep all test files in the same directory `sql/parser/testdata/`). It would also give us the freedom to move error cases in the specific statement files instead of a separate file.
</blockquote></details>

Hm I agree that would be nice, but it exceeds the time budget I'm willing to allocate to this PR. @rafiss can you file a followup issue for your team?


<!-- Sent from Reviewable.io -->

knz committed:
sql/parser: reveal broken statement telemetry

The SQL layer generates  statement telemetry in two phases:

- first the literal constants are removed statements to form app stat
  keys
- then the app stat keys are re-parsed into an AST
- the new AST is then re-pretty-printed using anonymization

The problem here is that there is no guarantee that the result of
stripping literals forms a valid SQL syntax. Moreover, it's possible
that the result of stripping the literals generates syntax that's
different from the original SQL semantically.

This patch exposes this difference. All the diffs in this commit
should be interpreted as bugs that cause invalid data in the telemetry.

Release note: None

knz commented:
@rafiss can you review the PR again and in particular take note of the 2nd commit. Then please itemize all the problems found in this PR as a comment inside #60722, or sub-issues. This needs to be picked up (probably incrementally) by multiple teams across the department.

rafiss reviewed (approved):
thank you! i've created an initial list and plan in #60722 

this looks great to me. you may need to update `pkg/sql/parser/BUILD.bazel` to let this datadriven test see the test data in the bazel sandbox. see for example https://github.com/cockroachdb/cockroach/blob/f5ed0e24804ead802a791d6ec858d8a59fe54aff/pkg/sql/logictest/BUILD.bazel#L60

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-:-MVw_oE4B4D--zejsF5V:b-7o9mvk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @knz, @otan, @RaduBerinde, @rafiss, and @solongordon)

___
*[pkg/sql/parser/parse_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60676#-MTr7ibG-E1ew3EA8z4S:-MVwaoSz3UPM3TzsiMrL:b-3onmn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdc023643ac3e2047a2458f240ee8a520d4c5a92/pkg/sql/parser/parse_test.go#L36)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Hm I agree that would be nice, but it exceeds the time budget I'm willing to allocate to this PR. @rafiss can you file a followup issue for your team?
</blockquote></details>

i've filed https://github.com/cockroachdb/cockroach/issues/62106


<!-- Sent from Reviewable.io -->

knz commented:
So in the end I used your trick to enclose the asterisks inside quotes to fix certain test failures.

However there is one that remains:
```
    --- FAIL: TestSerializesScheduledBackupExecutionArgs/multiple-tables-with-encryption-enterprise_user (0.04s)
        create_scheduled_backup_test.go:419:
                Error Trace:    create_scheduled_backup_test.go:419
                Error:          Should be true
                Test:           TestSerializesScheduledBackupExecutionArgs/multiple-tables-with-encryption-enterprise_user
                Messages:       could not find matching name for "BACKUP TABLE system.jobs, system.scheduled_jobs INTO 'nodelocal://0/backup' WITH encryption_passphrase = '*****', detached"
```

@miretskiy @pbardea @dt can you advise how to fix this?

knz commented:
Note: I did due diligence and fixed a number of broken backup/restore tests already. I am pretty confident this is not an obvious mistake, like me forgetting to update something trivial.

I have some reason to believe that there is serious confusion in the backup/restore (scheduling) code, that was really a latent bug, and that this PR is uncovering this bug/misdesign.

dt commented on pkg/sql/sem/tree/backup.go:
```diff
@@ -212,8 +212,12 @@ func (o *BackupOptions) Format(ctx *FmtCtx) {
-		ctx.WriteString("encryption_passphrase=")
-		o.EncryptionPassphrase.Format(ctx)
+		ctx.WriteString("encryption_passphrase = ")
+		if ctx.flags.HasFlags(FmtShowPasswords) {
```
my guess is this is what is tripping up the schedules code, specifically here: https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/create_scheduled_backup.go#L460 

with this change, since that call doesn't pass FmtShowPasswords, it is caught off-guard when tree.AsString doesn't round-trip the AST node back to the the same string

knz replied:
Thanks david, that fixes it indeed.

knz commented:
rebased and fixed the test error as per David's hint. If CI is happy I'll merge this.

knz commented:
I got a green CI run https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests/2856780?buildTab=dependencies

bors r=RaduBerinde,rafiss

knz commented:
@rafiss this probably needs to be back ported, at least the bug fixes in the first commit. Unfortunately my plate is a bit full for the coming weeks. Is someone on your team able to process this?
